Calling generic function from a generic funtion for a different generic type does not compile

### Describe the bug

The compiler fails when executing the following script:

```go
#!/usr/bin/env -S v run

struct Test {}

fn unmarshal[T]() ! {
	get_number[int]()!
}

fn get_number[T]() !T {
	return T(42)
}

unmarshal[Test]()!
```

The code doesn't make sense, but it's just a result of an extreme reduction to have a reproducible minimum. The original code looked like this:

```go
fn unmarshal_struct[T](mut typ T, a Any) ! {
	res := a.object()!
	...
	$for field in T.fields {
		...
			} $else $if field.typ is int {
				typ.$(field.name) = unmarshal_number[int](val)!
			...
```

When trying to execute the script above, the following error is printed on the command line:

```txt
❯ ./bug.vsh
==================
   ^~~~~~~~~~~~~~~~~~
/tmp/v_501/bug.13118805245557268757.tmp.c:17691:21: error: initializing '_result_main__Test' (aka 'struct _result_main__Test') with an expression of incompatible type '_result_int' (aka 'struct _result_int')
        _result_main__Test _t1 = main__get_number_T_int();
                           ^     ~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/bug.13118805245557268757.tmp.c:17698:4: warning: expression result unused [-Wunused-value]
  (*(main__Test*)_t1.data);
   ^~~~~~~~~~~~~~~~~~~~~~
2 warnings and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

It appears that the compiler "loves" the type `T` so much, so that it declares a local variable of the type `T` and assigns the result of `get_number` to it, although its result is of a different type.

### Expected Behavior

The compiler should generate code checking if `get_number` failed and propagate the error to the caller:

```go
fn unmarshal[T]() ! {
	get_number[int]()!
}
```

### Current Behavior

The compiler generates code assigning the result of `get_number` to a variable of the generic type `T` of `unmarshal`. The types differ:

```c
        _result_main__Test _t1 = main__get_number_T_int();
```

### Reproduction Steps

Save the `.vsh` script from the description above to a file, make it executable and execute it. You can also save it to a file `bug.v` file and execute `v bug.v`.

### Possible Solution

I wish I knew :-)

### Additional Information/Context

I suspect that the problem has something in common with error handling of the `Result` types. This is the important part of the generated C code:

```c
VV_LOCAL_SYMBOL _result_void main__unmarshal_T_main__Test(void) {
	_result_main__Test _t1 = main__get_number_T_int();
	if (_t1.is_error) {
		_result_void _t2;
		memcpy(&_t2, &_t1, sizeof(_result));
		return _t2;
	}
	
  (*(main__Test*)_t1.data);
	return (_result_void){0};
}

VV_LOCAL_SYMBOL _result_int main__get_number_T_int(void) {
	_result_int _t1;
	_result_ok(&(int[]) { ((int)(42)) }, (_result*)(&_t1), sizeof(int));
	return _t1;
}

VV_LOCAL_SYMBOL void main__main(void) {
	_result_void _t1 = main__unmarshal_T_main__Test();
	if (_t1.is_error) {
		panic_result_not_set(IError_name_table[_t1.err._typ]._method_msg(_t1.err._object));
	}
	
 ;
}
```

If I remove all exclamation marks to avoid error checks:

```diff
 struct Test {}

-fn unmarshal[T]() ! {
-	get_number[int]()!
+fn unmarshal[T]() {
+	get_number[int]()
 }

-fn get_number[T]() !T {
+fn get_number[T]() T {
 	return T(42)
 }

-unmarshal[Test]()!
+unmarshal[Test]()
```

The generated C code will be compiled without problem:

```c
VV_LOCAL_SYMBOL void main__unmarshal_T_main__Test(void) {
	main__get_number_T_int();
}

VV_LOCAL_SYMBOL int main__get_number_T_int(void) {
	int _t1 = ((int)(42));
	return _t1;
}

VV_LOCAL_SYMBOL void main__main(void) {
	main__unmarshal_T_main__Test();
}
```

Another way to get rid of the error is using the same generic type (`int`) for `unmarshal` as for `get_number`:

```diff
 	return T(42)
 }

-unmarshal[Test]()!
+unmarshal[int]()!
```

The generated code will use the same type-wrapping struct:

```c
VV_LOCAL_SYMBOL _result_void main__unmarshal_T_int(void) {
	_result_int _t1 = main__get_number_T_int();
	if (_t1.is_error) {
		_result_void _t2;
		memcpy(&_t2, &_t1, sizeof(_result));
		return _t2;
	}
	
  (*(int*)_t1.data);
	return (_result_void){0};
}

VV_LOCAL_SYMBOL _result_int main__get_number_T_int(void) {
	_result_int _t1;
	_result_ok(&(int[]) { ((int)(42)) }, (_result*)(&_t1), sizeof(int));
	return _t1;
}

VV_LOCAL_SYMBOL void main__main(void) {
	_result_void _t1 = main__unmarshal_T_int();
	if (_t1.is_error) {
		panic_result_not_set(IError_name_table[_t1.err._typ]._method_msg(_t1.err._object));
	}
	
 ;
}
```


### V version

V 0.3.4 43bc85d

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 1be798b.43bc85d
OS: macos, macOS, 13.3.1, 22E772610a
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

getwd: /Users/prantlf/Sources/v
vexe: /usr/local/Cellar/vlang/0.3.4/libexec/v
vexe mtime: 2023-05-27 16:01:02

vroot: OK, value: /usr/local/Cellar/vlang/0.3.4/libexec
VMODULES: OK, value: /Users/prantlf/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.40.1
Git vroot status: weekly.2023.21-27-g43bc85d3
.git/config present: true

CC version: Apple clang version 14.0.3 (clang-1403.0.22.14.1)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1355,11 +1355,9 @@ fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) ast.
 			}
 			return node.return_type
 		} else if typ := c.table.resolve_generic_to_concrete(func.return_type, func.generic_names,
-			concrete_types)
+			node.concrete_types)
 		{
-			if typ.has_flag(.generic) {
-				node.return_type = typ
-			}
+			node.return_type = typ
 			return typ
 		}
 	}

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1354,11 +1354,23 @@ fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) ast.
 				}
 			}
 			return node.return_type
-		} else if typ := c.table.resolve_generic_to_concrete(func.return_type, func.generic_names,
-			node.concrete_types)
-		{
-			node.return_type = typ
-			return typ
+		} else {
+			if node.concrete_types.len > 0 && !node.concrete_types.any(it.has_flag(.generic)) {
+				if typ := c.table.resolve_generic_to_concrete(func.return_type, func.generic_names,
+					node.concrete_types)
+				{
+					node.return_type = typ
+					return typ
+				}
+			}
+			if typ := c.table.resolve_generic_to_concrete(func.return_type, func.generic_names,
+				concrete_types)
+			{
+				if typ.has_flag(.generic) {
+					node.return_type = typ
+				}
+				return typ
+			}
 		}
 	}
 	return func.return_type

--- a/vlib/v/tests/generic_fn_with_nested_generic_fn_call_test.v
+++ b/vlib/v/tests/generic_fn_with_nested_generic_fn_call_test.v
@@ -1,0 +1,14 @@
+struct Test {}
+
+fn unmarshal[T]() ! {
+	get_number[int]()!
+}
+
+fn get_number[T]() !T {
+	return T(42)
+}
+
+fn test_generic_fn_with_nested_generic_fn_call() {
+	unmarshal[Test]()!
+	assert true
+}

